ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -47,16 +53,78 @@ var (
 	// headless svc will be of the form: {kuberay-cluster-name}-headless-worker-svc
 	headlessServiceSuffix = "headless-worker-svc"
 
-	// map of pod slices to workers in the slice
-	sliceToWorkers map[slice][]worker
-
 	// Flag arguments.
-	BindAddr   string
-	CACert     string
-	ServerCert string
-	ServerKey  string
+	BindAddr       string
+	CACert         string
+	KubeConfigPath string
+	ServerCert     string
+	ServerKey      string
 )
 
+func NewTPUWebhookServer(podLister listersv1.PodLister) *TPUWebhookServer {
+	return &TPUWebhookServer{
+		podLister: podLister,
+	}
+}
+
+func (t *TPUWebhookServer) Mutate(w http.ResponseWriter, r *http.Request) {
+	admissionReview := &admissionv1.AdmissionReview{}
+	if err := json.NewDecoder(r.Body).Decode(admissionReview); err != nil {
+		http.Error(w, "Error decoding request body", http.StatusBadRequest)
+		return
+	}
+
+	if admissionReview.Request.Kind.Kind == "Pod" {
+		klog.V(1).Info("Received review for Pod creation")
+		response, err := mutatePod(t.podLister, admissionReview)
```

Comment:
Since we're refactoring this already, I suggest adding a check as early as possible in the admission request to return 200 with no mutations if the pod doesn't request TPUs. I know we check containers requesting TPUs in `mutatePod`, but a separate check outside `mutatePods` seems safer and less prone to future bugs impacting non-TPU pods

Reply:
I think at the top of `mutatePod` is the earliest we can return an `AdmissionResponse` with `allowed: true`, since we first have to unmarshal the `admissionReview.Request.Object.Raw` (which happens at the top of `mutatePod`) and then iterate over the Pod containers to check for "google.com/tpu" limits. I could move the `extractPod` call to within `TPUWebhookServer.Mutate` but this is functionally the same as the current implementation.

Reply:
We now return early in `mutatePod` with [dce1bbe](https://github.com/GoogleCloudPlatform/ai-on-gke/pull/723/commits/dce1bbe4b09fd0a48304cb2d519414578a3ecdc0).